File: php/load_logic.php
<?php

 /**
  * logic_loader.php: directs form POST data to respective python scripts.
  *
  *   Note: either the 'print', or 'echo' statments are used to return values back
  *         to ajax scripts.  For example, to send data to an ajax script we may
  *         have the following lines in this php script:
  *
  *           print json_encode(array('key' => 'msg'));
  *
  *         The data can be accessed from the ajax script as follows:
  *
  *           console.log( data.key );
  *
  *   Note: to debug, or view the entire POST data array, enter the following in
  *         this php script:
  *
  *           print json_encode(array('post_array' => print_r($_POST,true)));
  *
  *         Again, the data can be accessed from the ajax script as follows:
  *
  *           console.log( data.post_array );
  *
  *   Note: performing multiple 'print json_encode( ... )' statements yields an
  *         illegal json syntax.  Specifically, it concatenates two, or more
  *         json objects. The receiving javascript file will fail overall for
  *         the ajax request, on the account of a 'Parse' error.  The receiving
  *         javascript is only allowed to receive one json representation, which
  *         may have nested json objects (not concatenated).  Therefore, only one
  *         such 'print' statement is allowed.
  *
  *   @json_encode( value ), returns the JSON representation / object of 'value'.
  */

// helper functions
  include(dirname(__FILE__) . '/helper.php');

// global variables
  $json = array();

// instantiate data / loader
  $obj_data   = new Obj_Data($_POST);
  $obj_loader = new Obj_Loader($obj_data);
  $obj_loader->logic_loader($json);

// redefine JSON, and return array to AJAX
  $json['json_creator'] = basename(__FILE__);

 /**
  * Class Obj_Loader: load proper SVM session
  */
  class Obj_Loader {

   /**
    * constructor: stores form data
    */
    public function __construct($form) {
      $this->form = $form;
    }

   /**
    * logic_loader(): receive the 'form_data' object, and determines the allocation
    *                 of its properties as parameters to respective python scripts.
    *
    * @json: 'reference' to the 'json' variable
    */
    public function logic_loader(&$json) {
    // HTML5 datalist supported: remove 'session_type'
      if ($this->form->datalist_support) {
        $session_type = $this->form->svm_session;
        unset($this->form->session_type);
      }
      else {
    // HTML5 datalist not supported: use counterpart, remove 'session_type'
        $session_type            = $this->form->session_type;
        $this->form->svm_session = $session_type;
        unset($this->form->session_type);
      }

      if ($session_type == 'training') {
      // Use HTML5 datalist fallback 'training_type'
        $this->form->svm_model_type   = $this->form->model_type;
        $this->form->svm_dataset_type = $this->form->dataset_type;
        unset($this->form->model_type);
        unset($this->form->dataset_type);

        $arr_result = array('result' => json_encode($this->form));
        $arr_result = array('data' => $arr_result);
        $result = shell_command('python ../python/svm_training.py', json_encode($arr_result));

        print json_encode($result);

      // Python returns JSON object
        if ( count((array)$result) > 0 ) {
          $json       = array_merge($json, array('msg_welcome' => 'Welcome to training'), $arr_result);
        }
      // Python returns nothing
        else {
          $json = array_merge($json, array('msg_welcome' => 'Welcome to training'));
        }
      }
      elseif ($session_type == 'analysis') {
      // Use HTML5 datalist fallback 'analysis_models'
        $this->form->svm_model_type = $this->form->model_type;
        unset($this->form->model_type);

        $arr_result = array('result' => json_encode($this->form));
        $arr_result = array('data' => $arr_result);
        $result     = shell_command('python ../python/svm_analysis.py', json_encode($arr_result));

      // Python returns JSON object
        if ( count((array)result) > 0 ) {
          $json       = array_merge($json, array('msg_welcome' => 'Welcome to analysis'), $arr_result);
        }
      // Python returns nothing
        else {
          $json = array_merge($json, array('msg_welcome' => 'Welcome to analysis'));
        }
      }
      else {
        print 'Error: ' . basename(__FILE__) . ', logic_loader()';
      }
    }
  }

?>

File: python/load_logic.py
#!/usr/bin/python

## @load_logic.py
#  This file allocates provided input to respective 'session_xxx_xxx.py' script,
#      and generates a return object as required.
#
#  Note: this script is executed from 'load_logic.php', using the 'exec( ... )'
#        equivalent method when implemented via the web-interface. Since the
#        web-interface is an AJAX process, the shelled into python script requires
#        print statements, when data is needed to be returned to the client-end.
#
#        The following will return the arguments passed into this AJAX shelled
#        into python script:
#
#            print sys.argv[1]
#
#        Then, in php, we can capture the returned data:
#
#            $output = shell_command("$command $parameters");
#            $arr_result = array('result' => $output);
#            $json = array_merge($json, $arr_result);
#
#        Which will allow us to return it to the client-end:
#
#            print json_encode($json);
#
#  Note: we can only call `print json_encode($json)` once. The receiving
#        javascript will interpret the data as follows:
#
#      console.log( data.result );
import sys, json
from session_data_new import Data_New
#from session_data_append import Data_Append
#from session_model_generate import Model_Generate
#from session_model_use import Model_Use

# local variables
list_error = []

if len(sys.argv) > 1:
  # determine if input is json decodable
  try:
    session_type = json.loads(sys.argv[1])['data']['settings']['svm_session']
  except Exception as e:
    error = 'Error: the provided \'svm_session\' is not json decodable, or not defined.'
    list_error.append(error)

  # redirect input to respective 'session_xxx_xxx.py' scripts
  if session_type == 'data_new':
    session = Data_New( sys.argv[1] )

    if session.check_arg_length():
      session.validate_svm_settings()

  elif session_type == 'data_append':
    Data_Append( sys.argv[1] )
  elif session_type == 'model_generate':
    Model_Generate( sys.argv[1] )
  elif session_type == 'model_use':
    Model_Use( sys.argv[1] )
  else:
    error = 'Error: the provided \'svm_session\' must be \'data_new\', \'data_append\', \'model_generate\', or \'model_use\'.'
    list_error.append(error)

  # return data
  if len(list_error) > 0:
    print json.dumps({ 'status': False, 'error': list_error }, sort_keys=True, indent=2, separators=(',', ': '))
  elif len(list_error) == 0:
    print json.dumps({ 'status': True, 'error': None })
